src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,351 @@
+        let sleep = TOKEN_VALID_DURATION;
+        tokio::time::advance(sleep).await;
+
+        // make sure this is the new token
+        let actual = cache.get_token().await.unwrap();
+        assert_eq!(actual, refresh);
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn expired_token_failure() {
+        let now = Instant::now();
+
+        let initial = Token {
+            token: "initial-token".to_string(),
+            token_type: "Bearer".to_string(),
+            expires_at: Some((now + TOKEN_VALID_DURATION).into_std()),
+            metadata: None,
+        };
+        let initial_clone = initial.clone();
+
+        let mut mock = MockTokenProvider::new();
+        mock.expect_get_token()
+            .times(1)
+            .return_once(|| Ok(initial_clone));
+
+        mock.expect_get_token()
+            .times(1)
+            .return_once(|| Err(CredentialError::non_retryable_from_str("fail")));
+
+        // fetch an initial token
+        let cache = TokenCache::new(mock);
+        let actual = cache.get_token().await.unwrap();
+        assert_eq!(actual, initial);
+
+        // wait long enough for the token to be expired
+        let sleep = TOKEN_VALID_DURATION;
+        tokio::time::advance(sleep).await;
+
+        // make sure we return the error, not the expired token
+        assert!(cache.get_token().await.is_err());
+    }
+
+    #[derive(Clone, Debug)]
+    struct FakeTokenProvider {
+        result: Result<Token>,
+        calls: Arc<std::sync::Mutex<i32>>,
+    }
+
+    impl FakeTokenProvider {
+        pub fn new(result: Result<Token>) -> Self {
+            FakeTokenProvider {
+                result,
+                calls: Arc::new(Mutex::new(0)),
+            }
+        }
+
+        pub fn calls(&self) -> i32 {
+            *self.calls.lock().unwrap()
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl TokenProvider for FakeTokenProvider {
+        async fn get_token(&self) -> Result<Token> {
+            // Release a token periodically. We give enough time for the
+            // waiters in a thundering herd to pile up.
+            tokio::time::sleep(Duration::from_millis(50)).await;
+
+            // Track how many calls were made to the inner token provider.
+            *self.calls.lock().unwrap() += 1;
+
+            // Return the result.
+            self.result.clone()
+        }
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+    async fn initial_token_thundering_herd_success() {
+        let token = Token {
+            token: "initial-token".to_string(),
+            token_type: "Bearer".to_string(),
+            expires_at: Some(Instant::now().into_std()),
+            metadata: None,
+        };
+
+        let tp = FakeTokenProvider::new(Ok(token.clone()));
+
+        let cache = TokenCache::new(tp.clone());
+
+        // Spawn N tasks, all asking for a token at once.
+        let tasks = (0..100)
```

Comment:
why cannot we have this thundering herd try to get a token that is already expired? Then atmost 1 refresh should happen?

Reply:
I think you are noticing that I had tests for  `initial_token_*` and `expired_token_*`, but I only have a unit test for `initial_thundering_herd_*`. There is no `expired_thundering_herd_*`.

The secret is that these are the same case, given how I wrote the code. (where `is_invalid(token) == true`), so I didn't want to write the same test again.

(We need both the `initial_token_*` and `expired_token_*` to actually test that `is_invalid(token) == initial || expired` .)

I changed the test name to be `invalid_token_thundering_herd_*`. (I will add `stale_token_thundering_herd_*` tests later).

Let me know if you were asking something else.